app/src/main/java/com/byoutline/kickmaterial/KickMaterialApp.kt

```diff
@@ -25,28 +39,35 @@ class KickMaterialApp : Application() {
         AndroidStubServer.start(this, NetworkType.UMTS)
         RetrofitHelper.setMsgDisplayer { msg -> this.showToast(msg, true) }
 
-        resetComponents()
+        initDagger()
     }
 
     @VisibleForTesting
-    @Synchronized
-    fun setComponents(mainComponent: GlobalComponent) {
-        component = mainComponent
-        Settings.set(debug = BuildConfig.DEBUG, containerViewId = R.id.container)
+    open fun initDagger() {
+        setComponents(createGlobalComponent())
     }
 
-    private fun resetComponents() {
-        val mainComponent = createGlobalComponent()
-        setComponents(mainComponent)
+    @VisibleForTesting
+    open fun setComponents(appComponent: AppComponent) {
+        component = appComponent
+        component.inject(this)
+        AppInjector.init(this)
+        SecretSauceSettings.set(debug = BuildConfig.DEBUG,
+                containerViewId = R.id.container,
+                bindingViewModelId = BR.viewModel,
+                viewModelFactoryProvider = ::getViewModelFactory)
```

Comment:
Application level init: Tells SecretSauce how to get ViewModelFactory. This allow usage of extension functions without needing to get ViewModelFactory in each Fragment/Activity